### ISSUE TITLE
Extend native import/export plugin to include additional entities

### ISSUE THREAD

**Issue by asmecher** [pkp/pkp-lib#3261]
Such as (probably not canonical)...

- Not-yet-published files
- Review assignments
- Discussions/queries
- Log entries
- ... etc

**Comment by asmecher** [comment]
(FYI, @jmacgreg)

**Comment by jmacgreg** [comment]
Oh, nice. CCing @mfelczak as well. Have we discussed the potential for using this as a PLN deposit option too? Should we ping Mark?

**Comment by asmecher** [comment]
The PLN plugin already uses the native import/export. This would extend its capabilities, at the risk of increasing storage requirements, and I could see us going either way (having the PLN plugin exclude the additional data and stick closer to its core mission, or allowing the additional data in to ease the "restoration" process). @mjordan, any preference?

**Comment by mjordan** [comment]
No objections here. If we don't want to load some of the new elements into an access version of a preserved journal, some preprocessing could strip them out. Or perhaps more sensibly, we preprocess the XML to be loaded to only include galleys, metadata, etc.

**Comment by diegoabadan** [comment]
See https://github.com/lepidus/fullJournalTransfer (for OJS 2.x)

**Comment by asmecher** [comment]
@diegoabadan, thanks, that's been a useful bit of work for OJS 2.x.

For 3.x, the XML import/export plugin uses the [filter framework](https://pkp.sfu.ca/wiki/index.php?title=Metadata_and_Filter_Framework), which essentially allows each entity to have a matching conversion filter from XML, and to XML. This will facilitate modular code -- the OJS 2.x XML import/export plugin was overly monolithic.

Do you have intentions to adapt any of the full journal transfer plugin to OJS 3.x? If so, we should combine efforts.

**Comment by mjordan** [comment]
Will this change apply to OJS 3 only, or also OJS 2? I'm asking because, having consulted with PKP Legal, we don't want to store anything that was never published in the PN or even on the staging server. Which means that the filtering out of the published stuff needs to happen within the OJS plugin.

**Comment by asmecher** [comment]
Yes, this would be OJS3. Filtering would basically involve stripping out the extra entities, either via XSL or a bit of PHP. (Having my hands in a bit of XSL these days, I'll heartily recommend PHP.)

**Comment by mjordan** [comment]
@asmecher would it make sense for Dimitris to add an integration point for this now, before we release the 3.x plugin? The callback (or whatever) to do the actual filtering could then be added later.

**Comment by asmecher** [comment]
I don't think so -- the plugin will get installed via the plugin gallery, and we can designate plugin version compatibility with each release, and updating the filter is essentially the same as updating the plugin (unless we're imagining hosting an .xsl e.g. on a PKP server, which would be a worse kind of dependency hell, I think).

**Comment by mjordan** [comment]
OK, thanks, I thought I'd ask in case it saved us some work and testing later. But just so I understand, the plugin will get installed in the Gallery initially, then included in the next release, correct? And it can be enabled either in the plugins list or directly in the Archiving tab. That probably doesn't matter but I just want to confirm that.

**Comment by asmecher** [comment]
I'd like to keep it in the Plugin Gallery throughout. We have 80 or so plugins already included in OJS and need to decrease that -- moving something to the plugin gallery is not an indication of lesser importance.

**Comment by diegoabadan** [comment]
Thanks for the info Alec!

It is not in our short term planning. We want to prioritize other plugins before.

**Comment by mjordan** [comment]
[This thread in the forums](https://forum.pkp.sfu.ca/t/is-ojs-gdpr-compliant/37521/20) is relevant. We probably should do the work to munge personal info before we release the 3.x version of the PN plugin.

**Comment by defstat** [comment]
@asmecher  for reviewRounds/ReviewAssignments

PRs
pkp-lib: pkp/pkp-lib#3722
ojs: pkp/ojs#1978

On the same branches, coming up Queries/QueryParticipants/Notes

**Comment by albig** [comment]
Dear @defstat, dear @asmecher, 

we've just discussed the same requirements on the OJS workshop in Heidelberg today. And I wonder, what's the current state of the this issue and the provided pull requests? Is this a good approach to follow and to contribute to?

**Comment by asmecher** [comment]
Hi @albig!

For the OJS/OMP 3.2 release (due within a few weeks), the only changes to the import/export XML will be 1) to support metadata versioning (https://github.com/pkp/pkp-lib/issues/4880), and 2) to add more options for including/referencing file content (https://github.com/pkp/pkp-lib/pull/5524).

I don't believe the PRs above have seen attention for a while, but @defstat can update you on the specifics. There have been some underlying changes (e.g. `SchemaDAO` and related code) that will need consideration, and the may also open some better opportunities for including entities in import/export XML.

Are there particular entities you're interested in including in the import/export XML?

**Comment by diegoabadan** [comment]
I believe that next week we can send a PR to resolve the #4195.

**Comment by asmecher** [comment]
That sounds great, @diegoabadan! We're planning to release OJS/OMP 3.2 at the end of February, and I can't promise we can get your PR included, but we'll try.

**Comment by KBodarwe** [comment]
Hello,
I was wondering what the status of the project is? As of 3.2 the feature does not seem to have made it into the release.
Thanks!

**Comment by asmecher** [comment]
@KBodarwe, this will be a bit of a long-term project, and we're hoping to achieve it by aligning the import/export tools with the API rather than trying to maintain parallel implementations. Is there a particular entity you're interested in supporting in the import/export toolset?

**Comment by SantiMaceri** [comment]
Hello @asmecher , i'm currently working on my thesis project, the main goal its to extend the native plugin so it can export (and import) more information than it does right now (like discussions, all the review stage, etc) for version 3.2.1-1.
I am analyzing the actual code and the xmls schemas so I can figurate out how to extend it.
Any documentation about how the plugin works  (especially on filters) will be helpful.
Thanks.

**Comment by asmecher** [comment]
Hi @SantiMaceri! Thanks for your offer -- we'd like to take you up on it, but before you start, I'd like to make sure we're also not digging into technical debt where we can avoid it. Is there a specific problem you need to solve with the import/export tools that requires more entities to be added?

**Comment by SantiMaceri** [comment]
Thanks for the quick answer @asmecher .
I must clarify that im working along side @gonetil, he's my thesis professor. 
We would like to give the chance to different journals to migrate a bigger part from a submission/article than they have right now.
To be more specific, we would like to add:

- The participants that work in the submission workflow

- All the discussions from all stages.

- The entire revision workflow

Currently,  i'm making progress with the participants. Next, it'd be the discussions, then the revision workflow.
Again, if you have any type of documentation that speeds up de project, it'd be fantastic.
And, if in any case you want to contact us to talk about the project, its also welcome. This is a thesis project, but we also want to contribute.

**Comment by dennmuel** [comment]
Hi all!

Sorry to just bump this issue, but I stumbled upon it while looking for a way to transfer a journal from one instance to another, ideally migrating more than "just" users, issues and articles (particularly the three entities mentioned by @SantiMaceri). Has there been progress "behind the scenes" or is there any other news about this?

Best regards
Dennis

**Comment by SantiMaceri** [comment]
Hi @dennmuel !
Yes! Along side @gonetil, we managed to extend the native plugin to export the three points that i've mentioned. (It's a work in progress). In fact, we're writing the thesis about this work right now.
In the following months we'are going to test it with portals from our university, and if it's ready we're going to upload it to a public repository.

Cheers
Santiago

**Comment by dennmuel** [comment]
Thanks for getting back so quickly, @SantiMaceri! Nice to hear it, please keep me posted. Until then: good luck for your thesis! :)

**Comment by defstat** [comment]
Hi @SantiMaceri! Is there a PR that we could test/review regarding that?

**Comment by SantiMaceri** [comment]
@defstat So sorry for the late answer! Not yet but we´re actually in the final steps of the thesis! When ready, i will public the repo so you can use it as you want.
Please hang on, we are doing our best and we hope it helps!

**Comment by KBodarwe** [comment]
Hey, sorry for pulling this up again. I'm working on a similar project (exporting articles including additional data for long term storage) and wanted to ask if there's any news in this?

@SantiMaceri is there any news on your project?

if not, @asmecher do you (or anyone else for that matter) have any hints on how to register your own filters? I have found the configureObject() function in the FilterDAO class, but have no idea on how or when to use it.

Thanks!
Best regards,
Karl

**Comment by gonetil** [comment]
Hi @KBodarwe ,
yes, there are some news. We have completed the first step of the project, and even though we were unable to test it in a production environment, all tests we run in a controlled env. resulted quite ok. 
It is important to note that both origin and destination OJS should be version 3.2 (we actually used ver. 3.2.1), and they must have the same configuration: sections, users and roles, review forms (it's all explain in the thesis)... Maybe someday the migration process includes all this settings, but so far we focused on submission mostly.

The thesis has been published in our institutional repository: http://sedici.unlp.edu.ar/handle/10915/120337 . It's in Spanish, I hope that's not a big issue for you.
Chapter 3 explains the solution we proposed, and chapter 4 shows all the tests we made. At the end of the document we also included the extensions we proposed to native.xsd to support new data (review rounds, review answers, review forms, discussions,...).
The code for the project is available under @SantiMaceri repostiory https://github.com/SantiMaceri/NativeXMLExtension

Let us know if this project is useful for you.
Best,
Gonzalo

**Comment by KBodarwe** [comment]
Hi @gonetil ,
that is actually very helpful, thanks. I'm not fluent in Spanish, but I can piece together the information from the code.

I see that you have extended the Native Import Export plugin. Did you add any new filters in the process? I'm trying to implement my project into a new plugin which needs new filters. Registering these into the database seems to be the problem.

**Comment by asmecher** [comment]
From PKP's side, we haven't made steps to improve the coverage of the XML, but our hope is that we can continue building out the REST API, adopting it from within the application as we do, then add some better command-line tools for interacting with the API outside the UI. The advantage of this is that we won't have to maintain conversion code for each entity to the import/export tools, and to the UI, and to the API -- we can ideally just do it once. That's not to say that we're deprecating the XML import/export toolset -- it'll be around for a long time -- but that we're not rushing to extend it to support more entities, as this generates a lot of brittle code the way it's currently implemented.

**Comment by KBodarwe** [comment]
Thanks @asmecher for the insight.
That seems like a long way still to go though.

If I may ask: I'm currently looking for a way to export on a per-article basis, including all submission files, comments, notes, additional files etc., into an XML format. What would be the best approach to this in your opinion? I've tried making a plugin similar to the native plugin, but can't find out when and how to register filters. I'm now trying a low-tech approach by going directly into the database (based on a shell script for example), but since submission data seems to be stored in a LOT of different tables, that sql statement could get real ugly real fast.
Any ideas for this problem? If this is the wrong place (which I'm semi-sure it is) for this question I'm happy to reformat into another place.

Thanks! 
Best regards, Karl.

**Comment by KBodarwe** [comment]
Hey @asmecher,
thanks very much for the info. I have started a thread in the PKP Forum:
https://forum.pkp.sfu.ca/t/ojs-3-3-registering-additional-filters-for-xml-export/70100

Cheers,
Karl.

**Comment by diegoabadan** [comment]
> Do you have intentions to adapt any of the full journal transfer plugin to OJS 3.x? If so, we should combine efforts.

Hi @asmecher,

A Brazilian university asked us to implement this plugin for OJS 3.3, so we're going to work on it.

Do you want to offer us guidance to better combine efforts?

**Comment by asmecher** [comment]
@diegoabadan, our long-term plans are still what's [described above](https://github.com/pkp/pkp-lib/issues/3261#issuecomment-932632544); the problem with the current import/export and filter frameworks is that supporting additional entities generates a **lot** of new code, and it's not maintainable. I'd be open to a PR that adds the support, but it's not code we'd be likely to support in the future. (Note that there have been a lot of changes between 3.3.0 and 3.4.0, so forward-porting code from 3.3.0 to 3.4.0 to `main` will also be a lot of work.)

**Comment by diegoabadan** [comment]
> our long-term plans are still what's https://github.com/pkp/pkp-lib/issues/3261#issuecomment-932632544;

I agree that it is a more sustainable solution. Our problem is that the university has a more immediate need for imports, which is why a shorter term solution was studied.

Note: other universities have consulted us with the same need.

> Note that there have been a lot of changes between 3.3.0 and 3.4.0, so forward-porting code from 3.3.0 to 3.4.0 to main will also be a lot of work.

Thanks! This was a concern, I had asked our team to evaluate it.

This university currently has an OJS with 20 journals, it is a typical user of LTS versions only.

**Comment by asmecher** [comment]
Relevant: https://forum.pkp.sfu.ca/t/native-xml-export-skipping-submission-files-in-ojs-3-4-0-x/82040

**Comment by diegoabadan** [comment]
We've released a preliminary package to start testing the OJS 3.3 plugin that extends what the native article/user import and export plugins do.

Development can be followed here:
https://github.com/lepidus/fullJournalTransfer

We're going to start more intensive testing now. Soon it should be ready for use in most cases.

**Comment by asmecher** [comment]
Thanks, @diegoabadan, that's likely to find a lot of interested users!

**Comment by diegoabadan** [comment]
The Federal University of São Paulo (Unifesp) confirmed that the OJS 3.3 plugin works after a journal migrated to its OJS. Our goal with this version has been achieved.

Additional features to the standard import and export of OJS users and articles:

    Journal data
    Navigation Menus
    Plugins Configs
    Sections
    Review Forms
    Review Assignments
    Review Rounds
    Review Files
    Reviewer Files
    Revision Files
    Stage Assignments
    Editor Decisions
    Discussions
    Metrics

**Comment by White-Horse-Press** [comment]
Sounds great, and just what we are looking for!
